libdebug/architectures/aarch64/aarch64_breakpoint_validator.py

```diff
@@ -0,0 +1,24 @@
+#
+# This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
+# Copyright (c) 2024 Roberto Alessandro Bertolini. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for details.
+#
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from libdebug.data.breakpoint import Breakpoint
+
+
+def validate_breakpoint_aarch64(bp: Breakpoint) -> None:
+    """Validate a hardware breakpoint for the AARCH64 architecture."""
+    if bp.condition.lower() not in ["r", "w", "rw", "x"]:
+        raise ValueError("Invalid condition for watchpoints. Supported conditions are 'r', 'w', 'rw', 'x'.")
+
+    if not (1 <= bp.length <= 8):
+        raise ValueError("Invalid length for watchpoints. Supported lengths are between 1 and 8.")
+
+    if bp.condition != "x" and bp.address & 0x7:
+        raise ValueError("Watchpoint address must be aligned to 8 bytes on aarch64. This is a kernel limitation.")
```

Comment:
I originally wanted to and actually started working on it, but I would personally advise against it for two reasons:
1. We don't actually have the proper infrastructure needed to support it right now, it would require a lot of tweaking and experimenting
2. What GDB does is better for human interaction, but for a consistent debugging script implementing this feature would cause a lot of unreliability and "unexplainable issues". Some unaligned watchpoints can be achieved with just one hardware watchpoint and some finessing, while others require two consecutive watchpoints. Scripts that "almost hit" the watchpoint number limit could then nondeterministically break on some runs and work flawlessly on others. I believe that forcing the user to manually split and align the watchpoints would save them the headache of understanding why some watchpoints are different than others, and why they don't have a consistent behavior.

aarch64 actually allows for fine-grained control over the breakpoint "shape", we might just have to wait for it to be integrated into the kernel (in the year 2085 maybe) to bring this feature.

Reply:
It seems legit, approved :)